clustermesh: policy: add --policy-default-local-cluster

MrFreezeex opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

Add a new option to select endpoint rules from the local cluster by default  for CNP, CCNP and k8s NP. It works in a similar way as namespaced NP (k8s NP and CNP). Enabling this new option will make sure that users do not add too wide network rules in ClusterMesh environments.

This new option is for now opt-in but meant to be a future new default once ready, that we provide proper tooling to identify (CC)NP that will be changed as a result and after a transition period.

Related to #36194 / https://github.com/cilium/cilium/issues/36194#issuecomment-2802430595

The main remaining point is about the tooling to identify which (CC)NP behavior will be changed when enabling this option. This could probably be added in a preflight container.

```release-note
policy: add a new option enable-default-restrict-local-cluster-policy to restrict policies rules to the local cluster in ClusterMesh environment
```

MrFreezeex commented:
/test

MrFreezeex commented:
/test

giorio94 reviewed (changes requested):
Thanks, the changes overall make sense to me. I've left a few suggestions inline, mostly minor. 

All clustermesh tests seem to have failed, so there's probably something broken somewhere. The connectivity suite should be also extended to correctly work with this new mode (unless it already does), so that we can also eventually switch a few matrix entries of the conformance tests to it.

MrFreezeex commented on pkg/option/config.go:
```diff
@@ -1374,6 +1374,10 @@ type DaemonConfig struct {
 	// ClusterID is the unique identifier of the cluster
 	ClusterID uint32
 
+	// EnableDefaultRestrictLocalClusterPolicy control whether policy rules assume
+	// by default the local cluster if not explicitly selected
+	EnableDefaultRestrictLocalClusterPolicy bool
+
```
Ah yep I double checked and it look like it's possible to do without this

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -3240,6 +3240,8 @@ clustermesh:
   enableEndpointSliceSynchronization: false
   # -- Enable Multi-Cluster Services API support
   enableMCSAPISupport: false
+  # -- Control whether policy rules assume by default the local cluster if not explicitly selected
+  enableDefaultRestrictLocalClusterPolicy: false
```
> it seems that policy options are mostly scattered

Yeah, I know. There is a place where you can find a few options (policyAuditMode, policyEnforcementMode, policyCIDRMatchMode) but no real grouping strictly speaking. 

> To me putting it close to clustermesh option could be more user friendly but I don't have a super strong opinion to not move it elsewhere though if it makes sense there too

Yes, I don't have a strong opinion either. From a certain point of view, this is fairly similar to the `maxConnectedClusters` option, that is somewhat specific to clustermesh, but actually affects the Cilium's behavior in a wider sense. Given that it ended up being being under the clustermesh stanza, it makes sense to keep this new one there as well.

aanm commented:
@MrFreezeex I've marked this PR as drafted. Feel free to mark it back as ready for review once you have addressed Marco's comments and resolved conflicts.

MrFreezeex committed:
clustermesh: policy: add --policy-default-local-cluster

Add a new option to select endpoint rules from the local cluster by default
(unless the rule select clusters explicitly) for CNP, CCNP and k8s NP. It works
in a similar way as namespace for CNP. Enabling this new option will make
sure that users do not add too wide network rules in ClusterMesh
environments.

This new option is for now opt-in but meant to be a future new default once
ready and after giving Cilium users a transition period.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

MrFreezeex commented:
> All clustermesh tests seem to have failed, so there's probably something broken somewhere.

The code changed a bit but AFAIK it was related to the fact that I forgot one to "remove" (/set to empty string) the cluster name in one call stack.

> The connectivity suite should be also extended to correctly work with this new mode (unless it already does), so that we can also eventually switch a few matrix entries of the conformance tests to it.

That's a good point I was assuming that it was working but not really... The current test deploys the "other-node" deployment on another cluster instead and assume that we can use the same exact test. Buut now to make it agnostic if it's on the same cluster I guess we would have to explicitly whitelist all clusters like so: https://github.com/MrFreezeex/cilium/blob/0dce0b05ca956ef9b3f34aee2b3dd7578afa2341/Documentation/network/clustermesh/policy.rst?plain=1#L70 (aka forcing the previous behavior whether the polocy default local cluster is enabled or not). If we were to do that do we need additional connectivity test specifically for this feature too? Also do you want to

MrFreezeex commented:
/test

MrFreezeex commented:
/test

MrFreezeex commented:
/test

giorio94 commented:
>> The connectivity suite should be also extended to correctly work with this new mode (unless it already does), so that we can also eventually switch a few matrix entries of the conformance tests to it.

> That's a good point I was assuming that it was working but not really... The current test deploys the "other-node" deployment on another cluster instead and assume that we can use the same exact test. Buut now to make it agnostic if it's on the same cluster I guess we would have to explicitly whitelist all clusters like so: https://github.com/MrFreezeex/cilium/blob/0dce0b05ca956ef9b3f34aee2b3dd7578afa2341/Documentation/network/clustermesh/policy.rst?plain=1#L70 (aka forcing the previous behavior whether the policy default local cluster is enabled or not). 

One compromise could be having a couple of policies which are templated with the expected cluster name (similarly to how it is already done in [this policy](https://github.com/cilium/cilium/blob/088f292c60a3ee49068adb9938e1308a8e1491e0/cilium-cli/connectivity/builder/manifests/client-egress-only-dns.yaml#L23) (referring to the local cluster in this case), while letting the others to simply use the wildcard for convenience, when necessary. WDYT?

> If we were to do that do we need additional connectivity test specifically for this feature too? 

Maybe one test that asserts that no cross-cluster traffic is allowed if the cluster name is not specified, or something like that, to be enabled only when the feature is enabled. I guess it also depends on whether it gets already automatically covered by any other test, when updated.

> Also do you see these update to cilium-cli as something that should go in this PR or it should be in a separate PR?

I don't personally have strong opinions on this, it looks fine to me to defer it to a separate PR considering that it would not affect Ci in its current form.

MrFreezeex commented:
> > > The connectivity suite should be also extended to correctly work with this new mode (unless it already does), so that we can also eventually switch a few matrix entries of the conformance tests to it.
> 
> > That's a good point I was assuming that it was working but not really... The current test deploys the "other-node" deployment on another cluster instead and assume that we can use the same exact test. Buut now to make it agnostic if it's on the same cluster I guess we would have to explicitly whitelist all clusters like so: https://github.com/MrFreezeex/cilium/blob/0dce0b05ca956ef9b3f34aee2b3dd7578afa2341/Documentation/network/clustermesh/policy.rst?plain=1#L70 (aka forcing the previous behavior whether the policy default local cluster is enabled or not).
> 
> One compromise could be having a couple of policies which are templated with the expected cluster name (similarly to how it is already done in [this policy](https://github.com/cilium/cilium/blob/088f292c60a3ee49068adb9938e1308a8e1491e0/cilium-cli/connectivity/builder/manifests/client-egress-only-dns.yaml#L23) (referring to the local cluster in this case), while letting the others to simply use the wildcard for convenience, when necessary. WDYT?
> 
> > If we were to do that do we need additional connectivity test specifically for this feature too?
> 
> Maybe one test that asserts that no cross-cluster traffic is allowed if the cluster name is not specified, or something like that, to be enabled only when the feature is enabled. I guess it also depends on whether it gets already automatically covered by any other test, when updated.

Sounds great, thanks for the insights!

> > Also do you see these update to cilium-cli as something that should go in this PR or it should be in a separate PR?
> 
> I don't personally have strong opinions on this, it looks fine to me to defer it to a separate PR considering that it would not affect Ci in its current form.

Ok great I am thus marking this PR ready for review and meanwhile will start doing that in another PR!

giorio94 commented on pkg/clustermesh/types/option.go:
```diff
@@ -135,3 +135,33 @@ func (_ QuirksConfig) Flags(flags *pflag.FlagSet) {
 	flags.MarkHidden("allow-unsafe-policy-skb-usage")
 }
+
+const PolicyAnyCluster = ""
+
+// PolicyConfig allows the user to configure config related to ClusterMesh and policies
+type PolicyConfig struct {
+	// PolicyDefaultLocalCluster control whether policy rules assume
+	// by default the local cluster if not explicitly selected
+	PolicyDefaultLocalCluster bool
+}
+
+var DefaultPolicyConfig = PolicyConfig{
+	PolicyDefaultLocalCluster: false,
+}
+
+func (_ PolicyConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool(
+		"policy-default-local-cluster", false,
+		"Control whether policy rules assume by default the local cluster if not explicitly selected",
+	)
+}
```
Maybe I wonder if these definitions should be directly moved to the policy package, given that not strictly clustermesh related.

MrFreezeex replied:
Since ClusterInfo is defined there and that it works mostly with that (except if done through the daemon legacy config :sweat_smile:), I figured it could be a good enough place but if anyone has a better idea I am happy to move this elsewhere ofc!

MrFreezeex commented:
/test

MrFreezeex commented:
/test

MrFreezeex commented:
/test

qmonnet commented on Documentation/network/clustermesh/policy.rst:
```diff
@@ -35,13 +35,38 @@ between two clusters. The cluster name refers to the name given via the
+    metadata:
+      name: "allow-cross-cluster-any"
+    spec:
+      description: "Allow x-wing deployed in the local cluster to contact rebel-base in any clusters"
```
```suggestion
      description: "Allow x-wing to be deployed in the local cluster to contact rebel-base in any cluster"
```

qmonnet reviewed (changes requested):
Looking at doc changes - I've got a couple of small nits, but looks good otherwise.

MrFreezeex commented:
/test

pippolo84 reviewed (approved):
Nice, thanks! :100: 

Just a non-blocking suggestion left inline.

MrFreezeex commented on pkg/clustermesh/types/option.go:
```diff
@@ -135,3 +135,33 @@ func (_ QuirksConfig) Flags(flags *pflag.FlagSet) {
+
+// PolicyLocalClusterName returns what should be considered the local cluster
+// name in network policies
+func (p PolicyConfig) PolicyLocalClusterName(localClusterName string) string {
```
Sounds reasonable, I did as you suggested, thanks :pray:

MrFreezeex commented:
/test
